Support for `serialize_as_any` runtime setting (#8830)
pydantic/pydantic: docs/concepts/serialization.md

### Serializing with duck-typing 🦆

!!! question "What is serialization with duck typing?"

    Duck-typing serialization is the behavior of serializing an object based on the fields present in the object itself,
    rather than the fields present in the schema of the object. This means that when an object is serialized, fields present in
    a subclass, but not in the original schema, will be included in the serialized output.

    This behavior was the default in Pydantic V1, but was changed in V2 to help ensure that you know precisely which
    fields would be included when serializing, even if subclasses get passed when instantiating the object. This helps
    prevent security risks when serializing subclasses with sensitive information, for example.

If you want to preserve the old duck-typing serialization behavior, you can use a runtime setting, or annotate individual types.

* Field / type level: use the `SerializeAsAny` annotation
* Runtime level: use the `serialize_as_any` flag when calling `model_dump()` or `model_dump_json()`

We discuss these options below in more detail:

#### `SerializeAsAny` annotation:

> If you want to preserve the old duck-typing serialization behavior, this can be done using the `SerializeAsAny` annotation on a type:

Reviewer: ```suggestion
If you want to duck-typing serialization behavior, this can be done using the `SerializeAsAny` annotation on a type:
```

which means that the model will ignore the schema and instead ask the object itself how it should be serialized.
In particular, this means that when model subclasses are serialized, fields present in the subclass but not in
the original schema will be included.

If `serialize_as_any` is set to `False` (which is the default), the model will be serialized using the schema,
which means that fields present in a subclass but not in the original schema will be ignored.

For example:

```py
from pydantic import BaseModel

class User(BaseModel):
    name: str

class UserLogin(User):
    password: str

class OuterModel(BaseModel):
    user1: User
    user2: User

user = UserLogin(name='pydantic', password='password')

print(OuterModel(user1=user, user2=user).model_dump(serialize_as_any=True)) # (1)!
"""
{
    'user1': {'name': 'pydantic', 'password': 'password'},
    'user2': {'name': 'pydantic', 'password': 'password'},
}
"""

print(OuterModel(user1=user, user2=user).model_dump(serialize_as_any=False)) # (2)!
"""
{
    'user1': {'name': 'pydantic'},
    'user2': {'name': 'pydantic'},
}
"""
```

1. With `serialize_as_any` set to `True`, the result matches that of V1.

> 2. With `serialize_as_any` set to `False` (the V2 default), fields present on the subclass,

Reviewer: I feel like it's maybe worth explaining the benefit of this, which may not be clear to users reading these docs — sometimes, you want to make sure that no matter what fields might have been added in subclasses, the serialized object will only have the fields listed in the original type definition. This can be useful if you might add something like `password: str` field in a subclass.

```py
from typing import List

from pydantic import BaseModel

class User(BaseModel):
    name: str
    friends: List['User']

class UserLogin(User):
    password: str

class OuterModel(BaseModel):
    user: User

user = UserLogin(name='samuel', password='pydantic-pw', friends=[UserLogin(name='sebastian', password='fastapi-pw', friends=[])])

print(OuterModel(user=user).model_dump(serialize_as_any=True)) # (1)!
"""
{
    'user': {
        'name': 'samuel',
        'friends': [
            {
                'name': 'sebastian',
                'friends': [],
                'password': 'fastapi-pw',
            }
        ],
        'password': 'pydantic-pw',
    }
}
"""

print(OuterModel(user=user).model_dump(serialize_as_any=False)) # (2)!
"""
{
    'user': {
        'name': 'samuel',
        'friends': [
            {
                'name': 'sebastian',
                'friends': [],
            }
        ],
    }
}
"""
```

1. Even nested `User` model instances are dumped with fields unique to `User` subclasses.
2. Even nested `User` model instances are dumped without fields unique to `User` subclasses.

> #### Overridng the `serialize_as_any` default (False)

Reviewer: ```suggestion
#### Overriding the `serialize_as_any` default (False)
```

> You can override the default setting for `serialize_as_any` at the project level by configuring a subclass of `BaseModel` that overrides the default for the `serialize_as_any` argument to `model_dump()` and `model_dump_json()`.

Reviewer: I think it's confusing that you say "at the project level" here, but then have `class User(BaseModel):` below, where it would not use the serialize as any behavior by default.

I would suggest changing the phrasing here to something like:
```suggestion
You can override the default setting for `serialize_as_any` by configuring a subclass of `BaseModel` that overrides the default for the `serialize_as_any` argument to `model_dump()` and `model_dump_json()`, and then use that as the base class (instead of `pydantic.BaseModel`) for any model you want to have this default behavior.
```

For example, you could do the following if you want to use duck-typing serialization by default:

```py
from typing import Any, Dict

from pydantic import BaseModel, SecretStr

class MyBaseModel(BaseModel):
    def model_dump(self, **kwargs) -> Dict[str, Any]::
        return super().model_dump(serialize_as_any=True, **kwargs)

    def model_dump_json(self, **kwargs) -> str:
        return super().model_dump_json(serialize_as_any=True, **kwargs)

> class User(BaseModel):

Reviewer: With my suggestion above, it's less important to change this, but either way I think it might make sense to have this be `class User(MyBaseModel):` just to emphasize that if you want that default everywhere, you'll want to have all your models inherit from the custom BaseModel subclass.
